File: client/tws_foundation_client/lib/src/services/models/outputs/update_output.dart
import 'package:csm_client/csm_client.dart';

/// {model} class for [UpdateOutput].
///
/// Defines a data model class that represents an {Output} object for an {Update} operation, storing
/// the result interest data from an {Update} operation.
final class UpdateOutput<T extends EntityI<T>> implements DecodableI, EncodableI {
  /// The new updated [T] instance object.
  late T updated;

  /// The original [T] instance, before the update operation. This property depends on the operation parameters and if there
  /// was an original [T] instance before the update operation with the same [EntityI.id] value.
  T? original;

  /// [decode] purposes [T] builder.
  final T Function() _entityBuilder;

  /// Creates a new [UpdateOutput] instance.
  UpdateOutput(this._entityBuilder) {
    updated = _entityBuilder();
  }

  @override
  void decode(DataMap encode) {
    final DataMap updatedData = encode.get('updated');
    updated = _entityBuilder();
    updated.decode(updatedData);

    final DataMap? originalData = encode.get('original', null);
    original = originalData == null ? null : _entityBuilder();
    original?.decode(originalData ?? <String, Object?>{});
  }

  @override
  DataMap encode() {
    return <String, Object?>{
      'updated': updated.encode(),
      'original': original?.encode(),
    };
  }
}

File: server/TWS Foundation/CSM Foundation/Database/Entity/Depot/IDepot_Update/UpdateOutput.cs
namespace CSM_Foundation.Database.Entity.Depot.IDepot_Update;

/// <summary>
///     [Record] for the output of an entity [Update] operation.
/// </summary>
/// <typeparam name="T"></typeparam>
public record UpdateOutput<T>
    where T : IEntity {
    /// <summary>
    ///     The new updated <typeparamref name="T"/> instance.
    /// </summary>
    public required T Updated { get; set; }

    /// <summary>
    ///     The original <typeparamref name="T"/> instance before the update operation.
    /// </summary>
    /// <remarks>
    ///     This property depends on the operation parameters and if there was an original entity before the update operation.
    /// </remarks>
    public T? Original { get; set; }
}
